routers/web/repo/issue.go

```diff
@@ -1636,6 +1638,27 @@ func ViewIssue(ctx *context.Context) {
 	ctx.Data["HasProjectsWritePermission"] = ctx.Repo.CanWrite(unit.TypeProjects)
 	ctx.Data["IsRepoAdmin"] = ctx.IsSigned && (ctx.Repo.IsAdmin() || ctx.User.IsAdmin)
 	ctx.Data["LockReasons"] = setting.Repository.Issue.LockReasons
+	var hiddenEvents *forms.UpdateCommentTypeForm
+	if ctx.IsSigned {
+		eventsRaw, err := user_model.GetSettings(ctx.User.ID, []string{"hidden_comment_types"})
+		if err != nil {
+			ctx.ServerError("GetSettings", err)
+			return
+		}
+		if eventsRaw["hidden_comment_types"] != nil && eventsRaw["hidden_comment_types"].SettingValue != "" {
+			err = json.Unmarshal([]byte(eventsRaw["hidden_comment_types"].SettingValue), &hiddenEvents)
```

Comment:
JSON encoding/decoding... I'm not sure if that's the most efficient and performant way to do this.

Given that these are a bunch of booleans, we can simplify it to use a bitset to store these bools, whereby each bit represent a comment type and not the comment number(e.x:
```
case 3:
		return f.Reference
	case 4:
		return f.Reference
	case 5:
		return f.Reference
```
While these are different commentTypes, they all have the same output, let's only store these outputs in the bitset.

I also see that the settings are stored as text, so because it's a binary format we could optionally create efficient itoa's and atoi's variant for binary numbers(given that viewing issues can be a popular task to do and thus want it to run efficient)

Reply:
^ It should be a BitField struct, such that the code can be re-used for in the future and not limited to this.

Reply:
> whereby each bit represent a comment type and not the comment number

I don't really understand this: these numbers are just the different comment types defined [here](https://github.com/go-gitea/gitea/blob/830ab75ce03e187ee7bff024ed8318db47a62f8b/models/issue_comment.go#L34-L110). 
The other idea with using such a bitset has the only problem that this struct is needed for the settings and this would just add extra code (I don't know if this would be much code, I don't know anything about bitsets and would have to look at this; if it's wrong what I thought about them just correct me).

Reply:
> I don't really understand this: these numbers are just the different comment types defined here.

Woops, my bad, I misunderstood it.

> The other idea with using such a bitset has the only problem that this struct is needed for the settings and this would just add extra code (I don't know if this would be much code, I don't know anything about bitsets and would have to look at this; if it's wrong what I thought about them just correct me).

Yeah that's correct, it will be additional code(200 LOC max) - but that can be justified, as JSON encoding/decoding has a bigger overhead than using the bitfield and in terms of saving/storing a string will be smaller with just storing numbers.

Reply:
OK, I tried to implement this now, but I have some understanding issues:
1. Probably a very bad question, but I don't really know what a BitField is and I didn't find something on the internet that is able to explain it :(... Would be very nice if you could explain a bit more what a BitField is or link something that could help me. Thanks!
2. I worked with bitsets until now, and I don't find a good way to convert it to a string that I can save in the db - do you know a way?

Sorry for my really bad questions, but I'm not a Go expert...

Reply:
The bitset/bitfield is just an `int64` which can be converted with the `strconv` package.

Reply:
Oh, OK thanks!

Reply:
> Probably a very bad question, but I don't really know what a BitField is and I didn't find something on the internet that is able to explain it :(... Would be very nice if you could explain a bit more what a BitField is or link something that could help me. Thanks!

Yeah what @KN4CK3R said, sorry for that. I meant "bitset"

Reply:
I pushed afc675c93 which uses bitsets instead of a JSON file